src/JuMPContainer.jl

```diff
@@ -183,13 +185,86 @@ Base.length(it::ValueIterator)  = length(it.x)
 
 type KeyIterator{JA<:JuMPArray}
     x::JA
+    dim::Int
+    next_k_cache::Array{Any,1}
+    function KeyIterator(d)
+        n = ndims(d.innerArray)
+        new(d, n, Array(Any, n+1))
+    end
+end
+
+KeyIterator{JA}(d::JA) = KeyIterator{JA}(d)
+
+immutable IndexableIndexsets end
+immutable NotIndexableIndexsets end
```

Comment:
I think that traits are cleaner for dispatch. I agree that now it isn't impacting performance because the dispatch of the function `next` is done using the type of k that is either a `Tuple` for the new iterator or an `Integer` for the old one. But if we manage to change the new iterator to use an `Integer` state as well, traits will be needed. I don't mind deleting them right now.